### ISSUE TITLE
Creating a kube-scheduler plugin on Microk8s

### ISSUE THREAD

**Issue by aNDREUET648** [kubernetes-sigs/scheduler-plugins#379]
#### Abstract
I'm trying to create a **kube-scheduler plugin on Microk8s**.
To do that I will implement my own scheduler image. I will also need to register, configure the plugin API, implement the extension point interfaces defined in the kubernetes scheduler framework package.
As plugins must be compiled with the scheduler, writing a wrapper allows to re-use the scheduler’s code in a clean way.
To do all of this I am using an article from medium [K8S - Creating a kube-scheduler plugin](https://medium.com/@juliorenner123/k8s-creating-a-kube-scheduler-plugin-8a826c486a1).
The article is truly fantastic and very didactic with a lot of documentation on the subject.

#### What is my problem?

The problem is that I need to reconfigure the vanilla scheduler and automate the way to deploy it (microk8s scheduler starts as an executable managed by systemd and not as a pod deployment) and the explanations provided by both Julio Renner's [article](https://medium.com/@juliorenner123/k8s-creating-a-kube-scheduler-plugin-8a826c486a1) and [kubernetes-sigs. Install Scheduler-plugins](https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/doc/install.md) are very poor and don't fit my needs.

#### Reproduction Steps

All the sections are very easy to follow until we get to the point near the end of the text [Deploying and using the Plugin](https://medium.com/@juliorenner123/k8s-creating-a-kube-scheduler-plugin-8a826c486a1):
- Build my scheduler container image and upload it to Docker Hub.
- I log into my master node (where the control plane is running).
- I make a backup copy of scheduler.yaml. In my case /var/snap/microk8s/current/args/kube-scheduler.
- Create a file /etc/kubernetes/networktraffic-config.yaml change the values according to your environment. **What do I do now?**
- Modify the file /etc/kubernetes/manifests/kube-scheduler.yaml to run  scheduler-plugins with Network Traffic. **And here what do I do**

From here I understand that you do something like ```kubectl apply -f kube-scheduler.yaml -n kube-system```

But as we already know this is not the way [Configuring MicroK8s services ](https://microk8s.io/docs/configuring-services)

I hope you can please help me to solve this situation that has been worrying me for a couple of weeks and I can't find the solution. Any help is appreciated.

Best Regards

Andreu

**Comment by Huang-Wei** [comment]
@aNDREUET648 not sure you have work the problem around, if not, here are some hints maybe helpful:

> Create a file /etc/kubernetes/networktraffic-config.yaml change the values according to your environment. What do I do now?

When you build your own scheduler with additional plugins, keep in mind that those delta plugins are NOT enabled by default. So this config file tells scheduler which plugins you want to enable, also you can disable some default plugins, or change some plugin's default arguments or weight.

> Modify the file /etc/kubernetes/manifests/kube-scheduler.yaml to run scheduler-plugins with Network Traffic. And here what do I do

This step is to modify the vanilla settings, like the vanilla scheduler image, along with passing & mounting the aforementioned config file.

**Comment by aNDREUET648** [comment]
![Captura de pantalla de 2022-08-22 22-43-47](https://user-images.githubusercontent.com/74447839/186361150-7d8c7c1e-263e-4fe3-b85c-094b7daf8ff0.png)
Hi @Huang-Wei, sorry to bother you I know you're a busy man.
I have tried to deploy a second scheduler on my MicroK8s cluster but I am still getting errors. Let me tell you:

- I have finished deploying a kube-scheduler with the SIG scheduler-plugins subproject and I have added the networktraffic plugin that julio renner detailed so well in his article on medium. I have had some difficulties since julio worked with a branch v0.19.8 and I implemented it with v0.22.6 with all the consequent changes since then in the framework. The 0.23 version of scheduler-plugins I tried to create a flat executable of the plugins that are there now (without the networktraffic plugin), but it fails in the make build ,I don't know why.
- On the other hand, MicroK8s runs as a systemd service the kubelite daemon which at the same time runs the subprocesses: scheduler, controller, proxy, kubelet, and the apiserver services. I find myself unable to replace the native scheduler.
- The option of a second scheduler has not been possible to get it, because almost all the existing documentation details the scheduler deployment as a static pod and not as a service of the operating system. Also the existing configurations of the scheduler as a systemd service are very vague or not detailed and in canonical despite being good people, they are not very communicative.
- The third option (the last one I try), is the deployment as a Helm Chart. Taking advantage that you already had a Helm Chart made, I tried to customize it with my built kube-scheduler and controller files. These, I have them in my docker registry (andreuet/kube-scheduler:latest and andreuet/controller:latest) to be called by the chart. Now the problem is that the liveness and readiness probes fail on any port (10259, 10251, 10248.....).
![Captura de pantalla de 2022-08-22 22-55-33](https://user-images.githubusercontent.com/74447839/186361008-9231b41f-d389-44ac-b090-b0848a0812c7.png)
I have tried, even increasing time.

On the other hand, the information included in the helm chart in

>  configmap.yaml

 is the following:

```
apiVersion: v1

kind: ConfigMap
metadata:
  name: scheduler-config
  namespace: {{ .Values.scheduler.namespace }}
data:
  scheduler-config.yaml: |
    apiVersion: kubescheduler.config.k8s.io/v1beta1
    kind: KubeSchedulerConfiguration
    leaderElection:
      leaderElect: false
    profiles:
    - schedulerName: scheduler-plugins-scheduler
      plugins:
        queueSort:
          enabled:
          - name: Coscheduling
          disabled:
          - name: "*"
        preFilter:
          enabled:
          - name: Coscheduling
        permit:
          enabled:
          - name: Coscheduling
        score:
          enabled:
          - name: NetworkTraffic
        reserve:
          enabled:
          - name: Coscheduling
        postBind:
          enabled:
          - name: Coscheduling
      pluginConfig:
      - name: Coscheduling
        args:
          permitWaitingTimeSeconds: 10
          deniedPGExpirationTimeSeconds: 3
      - name: NetworkTraffic
        args:
          prometheusAddress: "http://http://10.152.183.78:9090"
          networkInterface: "enx00e04c80471b"
          timeRangeInMinutes: 3
```

the deployment.yaml file is as follows:

```
kind: Deployment
apiVersion: apps/v1
metadata:
  name: {{ .Values.controller.name }}
  namespace: {{ .Values.controller.namespace }}
  labels:
    app: scheduler-plugins-controller
spec:
  replicas: {{ .Values.controller.replicaCount }}
  selector:
    matchLabels:
      app: scheduler-plugins-controller
  template:
    metadata:
      labels:
        app: scheduler-plugins-controller
    spec:
      serviceAccount: scheduler-plugins-controller
      containers:
      - name: scheduler-plugins-controller
        image: {{ .Values.controller.image }}
        imagePullPolicy: IfNotPresent
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    component: scheduler
  name: {{ .Values.scheduler.name }}
  namespace: {{ .Values.scheduler.namespace }}
spec:
  selector:
    matchLabels:
      component: scheduler
  replicas: {{ .Values.scheduler.replicaCount }}
  template:
    metadata:
      labels:
        component: scheduler
    spec:
      serviceAccountName: scheduler-plugins-scheduler
      containers:
      - command:
        - /bin/kube-scheduler
        - --address=0.0.0.0
        - --leader-elect=false
        - --config=/etc/kubernetes/scheduler-config.yaml
        - --scheduler-name=scheduler-plugins-scheduler
        image: {{ .Values.scheduler.image }}
        livenessProbe:
          httpGet:
            path: /healthz
            port: 10259
          initialDelaySeconds: 15
        name: scheduler-plugins-scheduler
        readinessProbe:
          httpGet:
            path: /healthz
            port: 10259
        resources:
          requests:
            cpu: '0.1'
        securityContext:
          privileged: false
        volumeMounts:
        - name: scheduler-config
          mountPath: /etc/kubernetes
          readOnly: true
      hostNetwork: false
      hostPID: false
      volumes:
      - name: scheduler-config
        configMap:
          name: scheduler-config
```

And finally the values.yaml file. The rest of the Helm Chart files have not been modified:

```
# Default values for scheduler-plugins-as-a-second-scheduler.
# This is a YAML-formatted file.
# Declare variables to be passed into your templates.

scheduler:
  name: scheduler-plugins-scheduler
  image: andreuet/kube-scheduler:latest
  namespace: scheduler-plugins
  replicaCount: 1

controller:
  name: scheduler-plugins-controller
  image: andreuet/controller:latest
  namespace: scheduler-plugins
  replicaCount: 1
```
I just hope to get there at some point.

Best regards

Andreu

**Comment by Huang-Wei** [comment]
To narrow down the issue, could you just run the vanilla install of Helm chart v0.23.10 and see if it works in microK8s?

**Comment by aNDREUET648** [comment]
Hello @Huang-Wei , I will test the Helm chart v0.23.10 on my microk8s v1.24 tonight and get back to you with what happened.  

Best regards

Andreu

**Comment by aNDREUET648** [comment]
Sorry to interrupt you again, but the problem is that if I use Helm chart version v0.23.10, the values.yaml file includes the plugins that I will use (just as it did before in the configmap.yaml file) but where do I include now ` profiles.pluginConfig.args` of the NetworkTraffic plugin? or the ones that the Coscheduling plugin had before?

`configmap.yaml v0.22.6`

```
......
      pluginConfig:
      - name: Coscheduling
        args:
          permitWaitingTimeSeconds: 10
          deniedPGExpirationTimeSeconds: 3
      - name: NetworkTraffic
        args:
          prometheusAddress: "http://http://10.152.183.78:9090"
          networkInterface: "enx00e04c80471b"
          timeRangeInMinutes: 3
```



The configmap.yaml file of the new version v0.23.10 does not have a configuration where to include the arguments (parameters) for the networktraffic plugin.

How to solve it?

Could this be my version of values.yaml in version v0.23.10?

```
#values.yaml v0.23.10 Helm Chart

# Default values for scheduler-plugins-as-a-second-scheduler.
# This is a YAML-formatted file.
# Declare variables to be passed into your templates.

scheduler:
  name: scheduler-plugins-scheduler
  ----->image: andreuet/kube-scheduler:latest
  namespace: scheduler-plugins
  replicaCount: 1

controller:
  name: scheduler-plugins-controller
  -----> image: andreuet/controller:latest
  namespace: scheduler-plugins
  replicaCount: 1

# LoadVariationRiskBalancing and TargetLoadPacking are not enabled by default
# as they need extra RBAC privileges on metrics.k8s.io.

plugins:
---->  enabled: ["Coscheduling","CapacityScheduling","NodeResourceTopologyMatch", "NodeResourcesAllocatable","NetworkTraffic"]

global:  
   # queueSort is not indented under extensions 
   # as it needs to be globally enabled once. 
  queueSort: ["Coscheduling"]
  extensions: 
    preFilter: ["Coscheduling", "CapacityScheduling"] 
    filter: ["NodeResourceTopologyMatch"] 
    postFilter: ["Coscheduling", "CapacityScheduling"] 
   ------> score: ["NodeResourceTopologyMatch", "NodeResourcesAllocatable","NetworkTraffic"] 
    permit: ["Coscheduling"] 
    reserve: ["Coscheduling", "CapacityScheduling"] 
    postBind: ["Coscheduling"]
```

Thank you for your concern

Andreu

**Comment by Huang-Wei** [comment]
@aNDREUET648 yup, the recent Helm chart refactoring PR #326 doesn't support pluginConfig's customization. And in recent refactoring to leverage v1beta3, this support isn't added either. I will create an issue to track on this.

Before this gets implemented, you have to hardcodde `as-a-second-scheduler/templates/configmap.yaml` with additional pluginConfig for your plugin. BTW: the values.yaml doesn't need the `global` section anymore - those fields are inferrable now with v1beta3.

And last thing:

> I will test the Helm chart v0.23.10 on my microk8s v1.24 tonight and get back to you with what happened.

Please use v1.23 to test with v0.23.10. There is no version skew support in the upstream among control-plane components (i.e., it may or may not work to test API Server v1.24 with scheduler v1.23).

**Comment by aNDREUET648** [comment]
Hi @Huang-Wei , due to a series of problems with version 1.23 I have had to downgrade to tag version v0.21.6 and it works with my microk8s version 1.24 (i.e. API server v1.24 works with the v1.21 scheduler).

The Helm chart of tag v0.21.6 is the last version that allowed to include args in `profiles.pluginConfig`

I don't know why, in version 1.23 when I generated the internal group files of the code generator with `./hack/update-codegen.sh`, it didn't generate the defaulters. In addition to the warnings generated by the final conversion function with the plugins `CoschedulingArgs `and `NodeResourceTopologyMatchArgs`.

Now with the tag version v0.21.6 I have been able to insert the args for my plugin (networktraffic) in addition to the configuration of the deployment.yaml that is defined in this version of the Helm Chart, both the `readinessProbe` and the `livenessProbe` work perfectly, and therefore, the scheduler works perfectly.

I appreciate your concern for the fact that you have opened an issue (**[Helm] support customizing pluginConfig #419**) to solve this problem with the parameterization of the plugins that brings me upside down.

Best Regards

Andreu
